pkg/dynamicaccess/controller.go

```diff
@@ -1,54 +1,210 @@
+// TODO: review return params: how to get back history ref ?
+func (c *controller) UploadHandler(
+	ctx context.Context,
+	refrefence swarm.Address,
+	publisher *ecdsa.PublicKey,
+	historyRootHash *swarm.Address,
+) (swarm.Address, swarm.Address, swarm.Address, error) {
+	ls := loadsave.New(c.getter, c.putter, requestPipelineFactory(ctx, c.putter, false, redundancy.NONE))
+	history, err := NewHistory(ls, historyRootHash)
+	if err != nil {
+		return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+	}
+	now := time.Now().Unix()
+	kvsRef, err := history.Lookup(ctx, now)
 	if err != nil {
-		return swarm.EmptyAddress, err
+		return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
 	}
-	if kvs == nil {
-		// new feed
-		// TODO: putter session to create kvs
-		kvs = kvsmock.New()
-		_, err = c.granteeManager.Publish(kvs, publisher)
+	kvs := kvs.New(ls, kvsRef)
+	historyRef := *historyRootHash
+	if kvsRef.Equal(swarm.ZeroAddress) {
+		err = c.accessLogic.AddPublisher(ctx, kvs, publisher)
 		if err != nil {
-			return swarm.EmptyAddress, err
+			return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+		}
+		kvsRef, err = kvs.Save(ctx)
+		if err != nil {
+			return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+		}
+		err = history.Add(ctx, kvsRef, &now)
+		if err != nil {
+			return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+		}
+		historyRef, err = history.Store(ctx)
+		if err != nil {
+			return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+		}
+	}
+	encryptedRef, err := c.accessLogic.EncryptRef(ctx, kvs, publisher, refrefence)
+	return kvsRef, historyRef, encryptedRef, err
+}
+
+func NewController(ctx context.Context, accessLogic ActLogic, getter storage.Getter, putter storage.Putter) Controller {
+	return &controller{
+		granteeList: nil, //NewGranteeList(ls, ps, swarm.EmptyAddress),
+		accessLogic: accessLogic,
+		getter:      getter,
+		putter:      putter,
+	}
+}
+
+func (c *controller) Grant(ctx context.Context, granteesAddress swarm.Address, grantee *ecdsa.PublicKey) error {
+	return c.granteeList.Add([]*ecdsa.PublicKey{grantee})
+}
+
+func (c *controller) Revoke(ctx context.Context, granteesAddress swarm.Address, grantee *ecdsa.PublicKey) error {
+	if !c.isRevokeFlagged(granteesAddress) {
+		c.setRevokeFlag(granteesAddress, true)
+	}
+	return c.granteeList.Remove([]*ecdsa.PublicKey{grantee})
+}
+
+func (c *controller) Commit(ctx context.Context, granteesAddress swarm.Address, actRootHash swarm.Address, publisher *ecdsa.PublicKey) (swarm.Address, swarm.Address, error) {
+	//HACKreplace mock with real kvs
+	var act kvs.KeyValueStore
+	if c.isRevokeFlagged(granteesAddress) {
+		act = kvsmock.New()
+		c.accessLogic.AddPublisher(ctx, act, publisher)
+	} else {
+		act = kvsmock.NewReference(actRootHash)
+	}
+
+	grantees := c.granteeList.Get()
+	for _, grantee := range grantees {
+		c.accessLogic.AddGrantee(ctx, act, publisher, grantee, nil)
```

Comment:
The error returned by `AddGrantee` is not checked.

Reply:
This will be solved later in a grantee related PR.